firestore/src/main/java/com/firebase/ui/firestore/paging/FirestorePagingAdapter.java

```diff
@@ -0,0 +1,146 @@
+
+    private final Observer<LoadingState> mStateObserver =
+            new Observer<LoadingState>() {
+                @Override
+                public void onChanged(@Nullable LoadingState state) {
+                    if (state == null) {
+                        return;
+                    }
+
+                    onLoadingStateChanged(state);
+                }
+            };
+
+    private final Observer<PagedList<DocumentSnapshot>> mDataObserver =
+            new Observer<PagedList<DocumentSnapshot>>() {
+                @Override
+                public void onChanged(@Nullable PagedList<DocumentSnapshot> snapshots) {
+                    if (snapshots == null) {
+                        return;
+                    }
+
+                    submitList(snapshots);
+                }
+            };
+
+    /**
+     * Construct a new FirestorePagingAdapter from the given {@link FirestorePagingOptions}.
+     */
+    public FirestorePagingAdapter(@NonNull FirestorePagingOptions<T> options) {
+        super(options.getDiffCallback());
+
+        mSnapshots = options.getData();
+
+        mLoadingState = Transformations.switchMap(mSnapshots,
+                new Function<PagedList<DocumentSnapshot>, LiveData<LoadingState>>() {
+                    @Override
+                    public LiveData<LoadingState> apply(PagedList<DocumentSnapshot> input) {
+                        FirestoreDataSource dataSource = (FirestoreDataSource) input.getDataSource();
+                        return dataSource.getLoadingState();
+                    }
+                });
+
+        mDataSource = Transformations.map(mSnapshots,
+                new Function<PagedList<DocumentSnapshot>, FirestoreDataSource>() {
+                    @Override
+                    public FirestoreDataSource apply(PagedList<DocumentSnapshot> input) {
+                        return (FirestoreDataSource) input.getDataSource();
+                    }
+                });
+
+        mParser = options.getParser();
+
+        if (options.getOwner() != null) {
+            options.getOwner().getLifecycle().addObserver(this);
+        }
+    }
+
+    /**
+     * If {@link #onLoadingStateChanged(LoadingState)} indicates error state, call this method
+     * to attempt to retry the most recent failure.
+     */
+    public void retry() {
+        FirestoreDataSource source = mDataSource.getValue();
+        if (source == null) {
+            Log.w(TAG, "Called retry() when FirestoreDataSource is null!");
+            return;
+        }
+
+        source.retry();
+    }
+
+    /**
+     * Start listening to paging / scrolling events and populating adapter data.
+     */
+    @OnLifecycleEvent(Lifecycle.Event.ON_START)
+    public void startListening() {
+        mSnapshots.observeForever(mDataObserver);
+        mLoadingState.observeForever(mStateObserver);
+    }
+
+    /**
+     * Unsubscribe from paging / scrolling events, no more data will be populated.
```

Comment:
Maybe `[...] no more data will be populated, but the existing data will remain.` since that's different from the other adapter.